parlai/tasks/taskmaster/agents.py

```diff
@@ -0,0 +1,290 @@
+            # Filter out single greet messages
+            if len(conversation) > 1:
+                self.ep_cheat_sheet[
+                    len(self.ep_cheat_sheet)
+                ] = tm_utils.gen_ep_cheatsheet(conversation)
+                convos_update += [conversation]
+        self.convos = convos_update
+
+    def num_examples(self):
+        ctr = 0
+        for ep in self.ep_cheat_sheet:
+            ctr += (
+                self.ep_cheat_sheet[ep][tm_utils.USER_NUM_EX]
+                + self.ep_cheat_sheet[ep][tm_utils.ASSIS_NUM_EX]
+            )
+        return ctr
+
+    def num_episodes(self):
+        # For two passes over the data: Once to teach USER and once to teach ASSISTANT
+        return len(self.convos) * 2
+
+    def get(self, episode_idx, entry_idx):
+        conversation = self.convos[episode_idx % len(self.convos)]
+        if episode_idx < len(self.convos):
+            # USER then ASSISTANT mode [First pass]
+            ep_done = (
+                entry_idx * 2 == self.ep_cheat_sheet[episode_idx][tm_utils.LAST_USER]
+            )
+            predecessor = conversation[entry_idx * 2]['text']
+            successor = conversation[entry_idx * 2 + 1]['text']
+        else:
+            # ASSISTANT then USER mode [Second pass]
+            ep_done = (
+                entry_idx * 2 + 1
+                == self.ep_cheat_sheet[episode_idx % len(self.convos)][
+                    tm_utils.LAST_ASSISTANT
+                ]
+            )
+            predecessor = conversation[entry_idx * 2 + 1]['text']
+            successor = conversation[entry_idx * 2 + 2]['text']
+
+        action = {
+            'id': self.id,
+            'text': predecessor,
+            'episode_done': ep_done,
+            'labels': [successor],
+        }
+
+        return action
+
+
+class WozDialogueTeacher(FixedDialogTeacher):
+    """
+    Teacher for spoken two-person dialogues with labels being responses for
+    the previous statement. The data is traversed twice (doubled), once for
+    modelling USER replies and once for modelling ASSISTANT replies.
+    """
+
+    def __init__(self, opt, shared=None):
+        opt['fn'] = "woz-dialogs.json"
+        super().__init__(opt)
+
+        if 'exclude-invalid-data' not in opt:
+            opt['exclude-invalid-data'] = True
+
+        if shared and 'convos' in shared:
+            # another instance was set up already, just reference its data
+            self.convos = shared['convos']
+            self.episode_map = shared['episode_map']
+            self.ep_cheat_sheet = shared['ep_cheat_sheet']
+            self.num_ex = shared['num_ex']
+        else:
+            # need to set up data from scratch
+            self.ep_cheat_sheet = {}  # Stores imp. info. about each episode
+
+            # Not all episodes have relevant examples for both USER and ASSISTANT
+            # episode_map keeps track of which episode index is useful for which speaker
```

Comment:
So say for example a conversation has the USER go first and then the ASSISTANT.
This conversation is useful when having USER as text and ASSISTANT as label but not the other way around since there is no reply to the ASSISTANT's statement. So we do not want to include this example when going over assistant pass. If we just index as usual, we might run into a situation where we might not be able to return anything in the action. So I create a map where keys are contiguous and values are actual conversation indexes in the data that are "usable". So for example 
convo 1:
U
A

convo 2:
U
A
U

So the first example useful when going over the ASSISTANT pass is actually convo 2
so episode_map["A"][0] = 1

Hope that makes sense?

Reply:
yes, i think i understand thanks